File: sdks/rails/lib/apollo_deploy_signal_sdk_rails/resources/projects.rb
# frozen_string_literal: true

require_relative "../transport"
require_relative "../errors"

module ApolloDeploySignalSdkRails
  # projects API domain
  class Projects
    attr_reader :transport

    def initialize(transport)
      @transport = transport
    end

    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [ProjectPageResponse]
    def list_projects()

      transport.request(
        method: :get,
        path: "/v1/projects",
        timeout_ms: nil,
      )
    end

    # @param project_id [String]
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [ProjectResponse]
    def get_project(project_id:)
      path_params = {
        "projectId" => project_id,
      }

      transport.request(
        method: :get,
        path: "/v1/projects/${projectId}",
        path_params: path_params,
        timeout_ms: nil,
      )
    end

    # @param project_id [String]
    # @param body [UpdateProjectRequest] request body
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [ProjectResponse]
    def update_project(project_id:, body:)
      path_params = {
        "projectId" => project_id,
      }

      transport.request(
        method: :patch,
        path: "/v1/projects/${projectId}",
        path_params: path_params,
        body: body,
        content_type: "application/json",
        timeout_ms: nil,
      )
    end

    # @param project_id [String]
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [EmailPageResponse]
    def list_emails(project_id:)
      path_params = {
        "projectId" => project_id,
      }

      transport.request(
        method: :get,
        path: "/v1/projects/${projectId}/emails",
        path_params: path_params,
        timeout_ms: nil,
      )
    end

    # @param project_id [String]
    # @param email_id [String]
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [EmailDetailResponse]
    def get_project_email(project_id:, email_id:)
      path_params = {
        "projectId" => project_id,
        "emailId" => email_id,
      }

      transport.request(
        method: :get,
        path: "/v1/projects/${projectId}/emails/${emailId}",
        path_params: path_params,
        timeout_ms: nil,
      )
    end

    # @param project_id [String]
    # @param email_id [String]
    # @raise [ApolloDeploySignalSdkRails::SDKError]
    # @return [EmailTimelineResponse]
    def get_email_timeline(project_id:, email_id:)
      path_params = {
        "projectId" => project_id,
        "emailId" => email_id,
      }

      transport.request(
        method: :get,
        path: "/v1/projects/${projectId}/emails/${emailId}/events",
        path_params: path_params,
        timeout_ms: nil,
      )
    end

  end
end

File: sdks/elixir/lib/apollo_signal/domains/projects.ex
defmodule ApolloSignal.Domains.Projects do
  @moduledoc "projects API operations."

  alias ApolloSignal.Client, as: Client
  alias ApolloSignal.Types, as: Types
  alias ApolloSignal.Transport, as: Transport
  alias ApolloSignal.Errors, as: Errors


  defstruct [:client]

  @type t() :: %__MODULE__{client: Client.t()}

  @spec new(Client.t()) :: t()
  def new(%Client{} = client), do: %__MODULE__{client: client}

  @spec list_projects(domain :: t(), options :: keyword()) :: {:ok, ApolloSignal.Types.ProjectPageResponse.t(), Transport.metadata()} | {:error, Errors.t()}
  def list_projects(%__MODULE__{} = domain, options \\ []) do
    path = "/v1/projects"

    request = %{
      method: :get,
      path: path,
      query: %{},
      headers: %{},
      cookies: %{},
      body: nil,
      content_type: nil,
      accept: "application/json",
      timeout_ms: Keyword.get(options, :timeout_ms, nil),
      max_retries: Keyword.get(options, :max_retries),
      retry_unsafe_requests: Keyword.get(options, :retry_unsafe_requests),
      idempotency_key: Keyword.get(options, :idempotency_key)
    }

    case Transport.request(domain.client.transport, request) do
      {:ok, response} ->
        {:ok, ApolloSignal.Types.from_project_page_response(Transport.json_body(response.body)), Transport.metadata(response)}

      {:error, %Errors{} = error} -> {:error, error}
    end
  end

  @spec get_project(domain :: t(), project_id :: String.t(), options :: keyword()) :: {:ok, ApolloSignal.Types.ProjectResponse.t(), Transport.metadata()} | {:error, Errors.t()}
  def get_project(%__MODULE__{} = domain, project_id, options \\ []) do
    path =
      Transport.path("/v1/projects/${projectId}", %{
        "projectId" => project_id
      })

    request = %{
      method: :get,
      path: path,
      query: %{},
      headers: %{},
      cookies: %{},
      body: nil,
      content_type: nil,
      accept: "application/json",
      timeout_ms: Keyword.get(options, :timeout_ms, nil),
      max_retries: Keyword.get(options, :max_retries),
      retry_unsafe_requests: Keyword.get(options, :retry_unsafe_requests),
      idempotency_key: Keyword.get(options, :idempotency_key)
    }

    case Transport.request(domain.client.transport, request) do
      {:ok, response} ->
        {:ok, ApolloSignal.Types.from_project_response(Transport.json_body(response.body)), Transport.metadata(response)}

      {:error, %Errors{} = error} -> {:error, error}
    end
  end

  @spec update_project(domain :: t(), project_id :: String.t(), body :: ApolloSignal.Types.UpdateProjectRequest.t(), options :: keyword()) :: {:ok, ApolloSignal.Types.ProjectResponse.t(), Transport.metadata()} | {:error, Errors.t()}
  def update_project(%__MODULE__{} = domain, project_id, body, options \\ []) do
    path =
      Transport.path("/v1/projects/${projectId}", %{
        "projectId" => project_id
      })

    request = %{
      method: :patch,
      path: path,
      query: %{},
      headers: %{},
      cookies: %{},
      body: Types.encode(body),
      content_type: "application/json",
      accept: "application/json",
      timeout_ms: Keyword.get(options, :timeout_ms, nil),
      max_retries: Keyword.get(options, :max_retries),
      retry_unsafe_requests: Keyword.get(options, :retry_unsafe_requests),
      idempotency_key: Keyword.get(options, :idempotency_key)
    }

    case Transport.request(domain.client.transport, request) do
      {:ok, response} ->
        {:ok, ApolloSignal.Types.from_project_response(Transport.json_body(response.body)), Transport.metadata(response)}

      {:error, %Errors{} = error} -> {:error, error}
    end
  end

  @spec list_emails(domain :: t(), project_id :: String.t(), options :: keyword()) :: {:ok, ApolloSignal.Types.EmailPageResponse.t(), Transport.metadata()} | {:error, Errors.t()}
  def list_emails(%__MODULE__{} = domain, project_id, options \\ []) do
    path =
      Transport.path("/v1/projects/${projectId}/emails", %{
        "projectId" => project_id
      })

    request = %{
      method: :get,
      path: path,
      query: %{},
      headers: %{},
      cookies: %{},
      body: nil,
      content_type: nil,
      accept: "application/json",
      timeout_ms: Keyword.get(options, :timeout_ms, nil),
      max_retries: Keyword.get(options, :max_retries),
      retry_unsafe_requests: Keyword.get(options, :retry_unsafe_requests),
      idempotency_key: Keyword.get(options, :idempotency_key)
    }

    case Transport.request(domain.client.transport, request) do
      {:ok, response} ->
        {:ok, ApolloSignal.Types.from_email_page_response(Transport.json_body(response.body)), Transport.metadata(response)}

      {:error, %Errors{} = error} -> {:error, error}
    end
  end

  @spec get_project_email(domain :: t(), project_id :: String.t(), email_id :: String.t(), options :: keyword()) :: {:ok, ApolloSignal.Types.EmailDetailResponse.t(), Transport.metadata()} | {:error, Errors.t()}
  def get_project_email(%__MODULE__{} = domain, project_id, email_id, options \\ []) do
    path =
      Transport.path("/v1/projects/${projectId}/emails/${emailId}", %{
        "projectId" => project_id,
        "emailId" => email_id
      })

    request = %{
      method: :get,
      path: path,
      query: %{},
      headers: %{},
      cookies: %{},
      body: nil,
      content_type: nil,
      accept: "application/json",
      timeout_ms: Keyword.get(options, :timeout_ms, nil),
      max_retries: Keyword.get(options, :max_retries),
      retry_unsafe_requests: Keyword.get(options, :retry_unsafe_requests),
      idempotency_key: Keyword.get(options, :idempotency_key)
    }

    case Transport.request(domain.client.transport, request) do
      {:ok, response} ->
        {:ok, ApolloSignal.Types.from_email_detail_response(Transport.json_body(response.body)), Transport.metadata(response)}

      {:error, %Errors{} = error} -> {:error, error}
    end
  end

  @spec get_email_timeline(domain :: t(), project_id :: String.t(), email_id :: String.t(), options :: keyword()) :: {:ok, ApolloSignal.Types.EmailTimelineResponse.t(), Transport.metadata()} | {:error, Errors.t()}
  def get_email_timeline(%__MODULE__{} = domain, project_id, email_id, options \\ []) do
    path =
      Transport.path("/v1/projects/${projectId}/emails/${emailId}/events", %{
        "projectId" => project_id,
        "emailId" => email_id
      })

    request = %{
      method: :get,
      path: path,
      query: %{},
      headers: %{},
      cookies: %{},
      body: nil,
      content_type: nil,
      accept: "application/json",
      timeout_ms: Keyword.get(options, :timeout_ms, nil),
      max_retries: Keyword.get(options, :max_retries),
      retry_unsafe_requests: Keyword.get(options, :retry_unsafe_requests),
      idempotency_key: Keyword.get(options, :idempotency_key)
    }

    case Transport.request(domain.client.transport, request) do
      {:ok, response} ->
        {:ok, ApolloSignal.Types.from_email_timeline_response(Transport.json_body(response.body)), Transport.metadata(response)}

      {:error, %Errors{} = error} -> {:error, error}
    end
  end

end
